crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -0,0 +1,326 @@
+//! # Visibility constraints
+//!
+//! During semantic index building, we collect visibility constraints for each binding and
+//! declaration. These constraints are then used during type-checking to determine the static
+//! visibility of a certain definition. This allows us to re-analyze control flow during type
+//! checking, potentially "hiding" some branches that we can statically determine to never be
+//! taken. Consider the following example first. We added implicit "unbound" definitions at the
+//! start of the scope. Note how visibility constraints can apply to bindings outside of the
+//! if-statement:
+//! ```py
+//! x = <unbound>  # not a live binding for the use of x below, shadowed by `x = 1`
+//! y = <unbound>  # visibility constraint: ~test
+//!
+//! x = 1  # visibility constraint: ~test
+//! if test:
+//!     x = 2  # visibility constraint: test
+//!
+//!     y = 2  # visibility constraint: test
+//!
+//! use(x)
+//! use(y)
+//! ```
+//! The static truthiness of the `test` condition can either be always-false, ambiguous, or
+//! always-true. Similarly, we have the same three options when evaluating a visibility constraint.
+//! This outcome determines the visibility of a definition: always-true means that the definition
+//! is definitely visible for a given use, always-false means that the definition is definitely
+//! not visible, and ambiguous means that we might see this definition or not. In the latter case,
+//! we need to consider both options during type inference and boundness analysis. For the example
+//! above, these are the possible type inference / boundness results for the uses of `x` and `y`:
+//!
+//!       | `test` truthiness | `~test` truthiness | type of `x`     | boundness of `y` |
+//!       |-------------------|--------------------|-----------------|------------------|
+//!       | always false      | always true        | `Literal[1]`    | unbound          |
+//!       | ambiguous         | ambiguous          | `Literal[1, 2]` | possibly unbound |
+//!       | always true       | always false       | `Literal[2]`    | bound            |
```

Comment:
Looks like Rust is trying to consider this a doctest (probably because of the indentation?) and failing to compile it.